webui/react/src/stores/experiments.tsx

```diff
@@ -0,0 +1,67 @@
+import { Map } from 'immutable';
+import React, { createContext, PropsWithChildren, useCallback, useContext, useState } from 'react';
+
+import { getExperiments } from 'services/api';
+import { V1Pagination } from 'services/api-ts-sdk';
+import { GetExperimentsParams } from 'services/types';
+import { ExperimentItem } from 'types';
+import handleError from 'utils/error';
+import { Loadable, Loaded, NotLoaded } from 'utils/loadable';
+
+type ExperimentPagination = {
+  experiments: ExperimentItem[];
+  pagination?: V1Pagination;
+};
+
+type ExperimentsContext = {
+  experimentsIndex: Map<string, ExperimentPagination>;
+  updateExperimentsIndex: (ws: Map<string, ExperimentPagination>) => void;
+};
+
+const ExperimentsContext = createContext<ExperimentsContext | null>(null);
+
+const encodeArgs = (args: { [key: string]: any }): string => {
+  const orderedKeys = Object.keys(args).sort();
+  return orderedKeys.map((key: string) => `${key}=${JSON.stringify(args[key])}`).join(',');
+};
+
+export const ExperimentsProvider: React.FC<PropsWithChildren> = ({ children }) => {
+  const [experimentsIndex, updateExperimentsIndex] = useState<Map<string, ExperimentPagination>>(
```

Comment:
Think these changes would make this more readable:
```
// above this part
type EncodedFilters = string

// then here
const [paginationCache, setPaginationCache] = useState<Map<EncodedFilters, PaginationInfo>>
```

Reply:
related, this seems like a better case for a `useRef`, since we are not apparently using it to trigger state updates-- I am assuming this: `updateExperimentsIndex(experimentsIndex.set(encodeArgs(filters), response));` doesnt trigger an update in any code that consumes it, do you know?

Reply:
This _does_ trigger updates, otherwise the components would never re-render when the data finishes fetching.

Reply:
well anyway it think would be more in line with idiomatic react (which we for the most part stick to in the codebase) to do this as an object literal and call the setter function `setExperimentsIndex`, then do 

`setExperimentsIndex((index) => ({...index, [encodeArgs(filters) : response})`

Reply:
@trentwatt worth noting that this is an immutable map so it has better performance characteristics than cloning object literals and also prevents accidental mutation.

Reply:
but also i think `experimentsIndex` is a confusing name and would prefer `experimentsCache` or something along those lines

Reply:
when we clone the object literals we don't actually copy the responses (values), just the references to them, so i dont see a big impact. 

i see a bigger risk from potentially behavior- rendering or otherwise- from doing things in a non idiomatic way like`updateExperimentsIndex(experimentsIndex.set(encodeArgs(filters), response));`

Reply:
if the map becomes huge because the user never refreshes the page, i think that is a problem regardless of whether we use `Map` or object literal... that maybe we should address

Reply:
@trentwatt Why do you think that code is non-idiomatic? Cloning is expensive because it must recreate the whole map not because it moves values.

Reply:
> Cloning is expensive because it must recreate the whole map 

I think I wrote my comment above while you were writing that, and it would be my response to it.

Reply:
> @trentwatt Why do you think that code is non-idiomatic?

unidiomatic to directly update the object returned by `setState`.

in this case i think it makes more sense to do `useRef`. we can still trigger state updates on relevant values by pulling them out.

Reply:
This is not directly updating the object returned by `setState`. `.set` doesn't modify the map, it returns a new map.

Reply:
it modifies the map and returns the modified map.

Reply:
I'm up to rename it `experimentsCache` or maybe it is `experimentsByParams`?